torch/fx/experimental/proxy_tensor.py

```diff
@@ -757,6 +759,106 @@ def disable_autocast_cache():
         torch.set_autocast_cache_enabled(old_value)
 
 
+class _ModuleStackTracer(PythonKeyTracer):
+    """Customized version of PythonKeyTracer that retains module stack
+    information in node.meta["nn_module_stack"].
+
+    FX symbolic trace actually does this already, but it relies on `self.root`
+    being the actual module being traced. Since make_fx traces a lambda of our
+    creation, things don't work properly.
+
+    So for this version we hold onto a reference to the original module
+    (scope_root) and monkeypatch it in when capturing the stack trace.
+
+    See Note [Preserving the nn module stack metadata during export non-strict mode]
+    """
+
+    def __init__(self, scope_root):
+        super().__init__()
+        self.scope_root = scope_root
+        self.proxy_paths = WeakKeyDictionary()
+        self.proxy_modules = WeakKeyDictionary()
+        self.counter = 0
+
+        self.module_id_cache = defaultdict(list)
+        for name, mod in self.scope_root.named_modules(remove_duplicate=False):
+            self.module_id_cache[id(mod)].append(name)
+
+        self_ = self
```

Comment:
Can you explain a bit what the strategy is? You mention we do monkeypatching, but I can't really tell what gets monkeypatched.

Reply:
I just copied the docstring from fbcode so not sure about the monkeypatching part, but the strategy here is that when we have a shared submodule (diamond shaped), this caching helps to always create the single stable nn_module_stack. I can delete the monkeypatching part.